公司一次断电之后，开发使用的mysql无法本地连接，通过Xshell连接到开发服务器上，启动mysql（service mysql restart）失败

查看mysql的log

```java
[ERROR] Can't open the mysql.plugin table. Please run mysql_upgrade to create it.

[ERROR] Fatal error: Can't open and lock privilege tables: Table 'mysql.user' doesn't exist

```

查看mysql配置文件

cat /etc/my.cnf

```java
[mysqld]
character_set_server=utf8mb4
init_connect='SET NAMES utf8mb4'

#

# Remove leading # and set to the amount of RAM for the most important data
# cache in MySQL. Start at 70% of total RAM for dedicated server, else 10%.
# innodb_buffer_pool_size = 128M
#
# Remove leading # to turn on a very important data integrity option: logging
# changes to the binary log between backups.
# log_bin
#
# Remove leading # to set options mainly useful for reporting servers.
# The server defaults are faster for transactions and fast SELECTs.
# Adjust sizes as needed, experiment to find the optimal values.
# join_buffer_size = 128M
# sort_buffer_size = 2M
# read_rnd_buffer_size = 2M
datadir=/var/lib/mysql
socket=/var/lib/mysql/mysql.sock




# Disabling symbolic-links is recommended to prevent assorted security risks
symbolic-links=0

# Recommended in standard MySQL setup
sql_mode=NO_ENGINE_SUBSTITUTION,STRICT_TRANS_TABLES

[mysqld_safe]
log-error=/var/log/mysqld.log
pid-file=/var/run/mysqld/mysqld.pid
```

可以看到数据存储位置为 /var/lib/mysql


从报错上来看，是由于缺少了mysql.plugin表以及mysql.user表所致的，不过现在连启动都启动不了，
进入到/var/lib/mysql 中查看mysql目录下确实没有以上两个表

<<<<<<< HEAD
最后在网上找到执行以下命令可以解决

mysql_install_db --user=mysql --basedir=/usr/ --ldata=/var/lib/mysql/

执行命令之后在/var/lib/mysql/mysql/ 目录下新建了所需的那几张表

```java
-rw-rw----. 1 mysql mysql     8586 Sep 29 17:01 plugin.frm
-rw-rw----. 1 mysql mysql        0 Sep 29 17:01 plugin.MYD
-rw-rw----. 1 mysql mysql     1024 Sep 29 17:01 plugin.MYI
-rw-rw----. 1 mysql mysql     9996 Sep 29 17:01 proc.frm
-rw-rw----. 1 mysql mysql        0 Sep 29 17:01 proc.MYD
-rw-rw----. 1 mysql mysql     2048 Sep 29 17:01 proc.MYI
-rw-rw----. 1 mysql mysql     8875 Sep 29 17:01 procs_priv.frm
-rw-rw----. 1 mysql mysql        0 Sep 29 17:01 procs_priv.MYD
-rw-rw----. 1 mysql mysql     4096 Sep 29 17:01 procs_priv.MYI
-rw-rw----. 1 mysql mysql     8800 Sep 29 17:01 proxies_priv.frm
-rw-rw----. 1 mysql mysql     1386 Sep 29 17:01 proxies_priv.MYD
-rw-rw----. 1 mysql mysql     5120 Sep 29 17:01 proxies_priv.MYI
-rw-rw----. 1 mysql mysql    10684 Sep 29 17:01 user.frm
-rw-rw----. 1 mysql mysql      676 Sep 29 17:13 user.MYD
-rw-rw----. 1 mysql mysql     2048 Sep 29 17:16 user.MYI
```

启动mysql,可以成功了

但是访问权限拒绝了

需要重新在user表中新增登录的密码

update user set password=password('xxxxxxxx') where user='root';

最后再开启远程登录

方法一：

GRANT ALL PRIVILEGES ON *.* TO 'root'@'%' IDENTIFIED BY 'P365db0302' WITH GRANT OPTION;?
FLUSH PRIVILEGES;?

方法二：

mysql>use mysql;

mysql>update user set host = '%' where user = 'root';

mysql>select host, user from user;

重启mysql 

service mysql restart

使用navicat 已经可以连接到mysql了
=======
>>>>>>> parent of ff321d8... Update 一次数据库无法连接问题解决.md
